### DISCUSSION TITLE
collaborative quarto editing

### DISCUSSION THREAD

**Discussion by jmbuhr** [quarto-dev/quarto-cli#405] [Feature Requests]
What are your thoughts on collaboration with quarto?

For academic writing, I think the main "competitor" to match in terms of features is https://www.overleaf.com/ but quarto has the potential for a way smoother workflow, with `preview` for html output being much faster than recompiling on overleaf.
The main hurdle so far is collaboration with people who are focused on the content, who don't have to run the code but just want to change a couple of words, add a sentence, remove a section or highlight a piece and add a comment and have these changes tracked without having to know git or manually commit changes.

I assume these things fit in nicely with RStudio's professional products, but when the computations are not at the center, which is often the case when working on a manuscript, the added price would likely make people stick with e.g. overleaf, especially since in academia it is harder to spend money on such tooling compared to industry.

So, tldr. I would be over the moon to eventually see a minimal collaborative editing solution for quarto documents with a super low barrier to entry, like the google docs of scientific writing.

**Comment by jjallaire** [comment]
Yes, we 100% see these same issues and its definitely on our roadmap to address.

**Reply by john-h-carson-jr** [reply]
That would be extraordinarily helpful!!!

**Reply by thomasjgrady** [reply]
Also agree with the usefulness. Support for [critic markup](https://fletcher.github.io/MultiMarkdown-6/syntax/critic.html) in particular would be great.

Our group SLIM @ Georgia Tech has been using a quarto-like workflow for quite sometime based of a [fork](https://github.com/slimgroup/ScholarlyMarkdown) of ScholarlyMarkdown that provides many (but obviously not all!) of the features quarto does. In particular it supports critic markup, which we have found very useful. It may be a good starting point for adding these features.

**Reply by sebastian-gerdes** [reply]
Hello! 

I am very interested in such as a solution as well. 

I have a suggestions for a new R package (that I might write if I find the time) and would like to get some feedback:

The idea for collaboration between two persons (A and B) would be along the following lines:
A starts doc_01.qmd
B edits doc.qmd and saves it to doc_02.qmd

Then invoking `wordwise('doc_01.qmd', 'doc_02.qmd') should invoke an interactive process and write to 'doc_03.qmd', in which each change of B should be either:
- accepted
- rejected
- postponed for later decision
- marked for manual edit

I think it would facilitate some nice workflows - for example I could give the *.qmd file to my collaborator / supervisor, he /she could propose some changes in the *.qmd-file and he /she wouldn't have to get in the markdown too much if he / she doesn't want and I could incorporate the proposed changes.

What do you think about such an approach?

**Reply by mcanouil** [reply]
FYI, this is the idea of <https://diffviewer.r-lib.org> to compare testthat/vdiffr unit tests snapshots.

**Reply by sebastian-gerdes** [reply]
Thanks for the pointer! The diffviewer has a lot of the features that I am thinking about, however, as far as I can see no "structured" process for accepting / rejecting proposed changes. I will have a closer look at the diffviewer package soon - maybe I can build something on top!

**Reply by jmbuhr** [reply]
No need to re-invent the wheel, this is exactly what git (and GitHub) is for. This structured process you are looking for is just a Pull Request with a Review. And if you have a collaborator that you can't get to use git, they are also unlikely to want to use an additional R package.

**Reply by jameshowison** [reply]
I 'down-thumbed' a comment about using git/GitHub for this purpose.  Felt rude, so adding a comment to explain. 

I use things like google docs for collaborative writing which is occurring simultaneously, often while we are in the same room. I enjoy seeing other people's writing as it happens, being able to gather and refactor content quickly.  git and GitHub don't offer that opportunity.

That said, Overleaf connected to GitHub is great for long term tracking, so as I see it collaborative editors and version control systems are different technologies with related usecases but quite different timeframes.

**Reply by jmbuhr** [reply]
For reference, I'm also the person that posted the original feature request, where I ask for exactly what you descibe ;)

As such, my comment above does not imply that git can replace this collaborative, real-time editing. But what it already replaces is e.g. an R package that would do the same job. Anything that doesn't add the convenience of something like overleaf / google docs and their real-time nature, but still adds it's own barrier to entry is most likely better off simply being git.

**Reply by jimjam-slam** [reply]
For real-time collaboration, VSCode's [Live Share](https://visualstudio.microsoft.com/services/live-share) can also be a really nice option. The pros I really love are:

* You can see folks' cursors moving around in real-time
* You can follow someone else's cursor as they navigate a file (or switch files)
* You can optionally grant collaborators the right to run terminal commands if you're exploring
* Conversely, you can make the entire session read-only if you just want to show someone around a codebase
* You can share open web servers with collaborators, so they can test a Shiny app or a read a rendered Quarto doc alongside you (this happens automatically by default, I think, but you can change that if you don't like it)
* VSCode will actually add a "Co-authored by:" message with collaborators' names and email addresses to the end of commit messages when you commit while Live Sharing
* If collaborators don't have VSCode installed, they can connect to the Live Share session through a VSCode browser instance

![Animation of Live Share from Microsoft's site, showing two people engaged in a Live Share session](https://visualstudio.microsoft.com/wp-content/uploads/2023/01/v2-Edit-Comp_FINAL-optimized750-1.gif)

**Reply by jimjam-slam** [reply]
One thing missing from Live Share is that someone actively has to start a session and others have to join it—it would be nice if something could alert other collaborators when they open a repo that someone else is also working on it! Maybe there's a VSCode extension in that...

**Reply by kinh-nguyen** [reply]
@jimjam-slam do you have experience on keeping the session live for a long time? I think normally others won't join right away, and the session has been very short-lived for me.

Another option is write quarto in a Jupyter notebook which has many options to collaborate (e.g., Google Colab). I haven't checked if the trackchange feature is there or not.

**Reply by jimjam-slam** [reply]
@kklot I'm not sure, tbh! I've had sessions that _lasted _ a long time, but if you're talking about ones where the host has been idle for a long time before anyone joined, I haven't tried that before.
L

**Comment by john-h-carson-jr** [comment]
That does sound interesting.

**Reply by sebastian-gerdes** [reply]
Thanks for the feedback! I will continue to think about the project and I might post updates here, so people could test it!

**Comment by grasshoppermouse** [comment]
Just use google docs. It works great for this purpose.

**Reply by harrelfe** [reply]
Google docs and (for markdown .md files) the Microsoft Sharepoint in-line text editor have perhaps the lowest barriers to participation of multiple authors of different skill sets.

**Reply by jameshowison** [reply]
Trouble I run into is that people write in google docs and instantly start using the layout tools, breaking markdown instantly.  Care to share anything about the workflow that works for you?  I'm googling "constrain google docs to plain text only" but not seeing anything promising :)

**Reply by john-h-carson-jr** [reply]
Let me suggest the following workflow, which includes downloading to Word and using PanDoc:

GoogleDoc (collaboration, possibly including format) ==> Word (Download as .docx) ==> Markdown (via pandoc) ==> replace generated content with code chunks ==> save as .qmd

**Reply by mcanouil** [reply]
Or use an add-on such as https://workspace.google.com/marketplace/app/docs_to_markdown/700168918607

**Comment by lamurian** [comment]
I stumbled on this package by chance, have you ever checked this one? This supports collaborative editing of qmd/rmd/rnw by leveraging Google Docs.

https://claudiozandonella.github.io/trackdown/

**Reply by jjallaire** [reply]
Yes, that is an excellent approach indeed!

**Reply by TS-CUBED** [reply]
Hmm, this looks like an interesting approach, but something that doesn't need R would be nice.

**Comment by hlageek** [comment]
The best solution I have seen so far are the online markdown editors. [https://hackmd.io](https://hackmd.io) checks many boxes (renders markdown, real-time collab, comments possible on the preview, links for filesharing). Importantly, it can be integrated with GitHub, so changes in the narrative text can be merged back into a project repo and `qmd` can be subsequently fully re-rendered. There is a self-hosted fork [https://github.com/hedgedoc/hedgedoc](https://github.com/hedgedoc/hedgedoc). The only downside is for the project maintainer who has to push changes from HackMD into remote (and handle conflicts), but there are basically no barriers for the collaborators. Unfortunately, so far no support for [critic markup](https://fletcher.github.io/MultiMarkdown-6/syntax/critic.html). 

IMHO, a version of HackMD with critic markup support and some sort of cache for rendering the latest code chunk outputs would be pretty close to the Holy Grail of collaborative quarto editing.

I am also eyeing [https://github.com/typst/webapp-issues](https://github.com/typst/webapp-issues), as `quarto` already supports `typst`. If their web app somehow supports quarto in turn, it could be an interesting overleaf alternative.

**Comment by da5nsy** [comment]
Flagging a parallel discussion I just started on the jupyter-book discussion section: 
- https://github.com/orgs/jupyter-book/discussions/2161

**Comment by Arshiyasan** [comment]
Have there been any updates on this? My current workflow is local editting -> pushing to github -> an online markdown editor -> pulling. Would love to know what simple/robust workflows others are using!

**Comment by petertanksley** [comment]
Linking this here, Quartorium could be an interesting approach!

https://github.com/quarto-dev/quarto-cli/discussions/12968
https://github.com/andjar/quartorium

**Reply by da5nsy** [reply]
Very cool! As I understand it - there isn't a hosted version of this anywhere yet, right? (which people could use via the web)

**Reply by MichelNivard** [reply]
I opnce build something similar: https://github.com/MichelNivard/resolve 

Its hosted at resolve.pub, but no one ever uses it. The biggest issue I guess was that I stored the quarto doc in a .ipynb file, which because of its structure I found easier to parse from and too, that might have been too big of a nucance?

THis is a showcase: https://www.youtube.com/watch?v=uBmBZ4xLeys&t=1s

I might re-visist or reach out to the quartorium author to see if I can contribute.

**Reply by MichelNivard** [reply]
I think across his and my solution I see where the friction arises: quartorium renders the .qmd on the fly, which is an issue for a few reasons 1. code execution opens up security issues 2. data needs to be accesible/travels 3. hard to ensure dependencies are in place and project environment control becomes very hard. I tiried to get around that by editign a rendered document, and coming up with an interim file format which holds the qauarto doc and the wysiwyg, it was very in-enlegant. IMO it will be very hard to "have it both ways" where a document retains its status as pre render quarto file for the technical authors, but also is a WYSIWYG word like with computation/code abstracted out and the computational results previewable (i.e. (pre)rendered).

**Reply by tockudex** [reply]
Would it make sense to, say, fork Hedgedoc and integrate Quarto and Latex/Tpyst/etc. specific features from your https://github.com/andjar/quartorium or your previous work @MichelNivard into it?

**Comment by agbocsardi** [comment]
In case anybody is interested, there is a way to use [Overleaf Github Sync](https://docs.overleaf.com/integrations-and-add-ons/git-integration-and-github-synchronization/github) to allow academic collaborators to stay in comfortable waters. There is a caveat that you need to delete and recreate the qmd files using the Overleaf UI to be allowed to edit them, but to me that extra copy-paste step was worth it.

I wrote the workflow up in a bit more detail [on my blog here](https://agbocsardi.com/posts/quarto-meets-overleaf/) in case anybody needs a bit more direction.

**Comment by jameshowison** [comment]
Very cool, glad you shared @agbocsardi 

My fervent hope is that quarto peeps and Overleaf to make quarto a full class citizen at Overleaf. Perhaps something with `execute:
  freeze: true` would help manage the computation costs?

**Comment by harrelfe** [comment]
The comment about computation costs reminded me to mention that whatever approach is chosen one needs to be able to install CRAN R packages on the fly.

**Comment by fBedecarrats** [comment]
Hello, I am searching again on this and stumble upon this page: https://deepwiki.com/quarto-dev/quarto/4.2-collaborative-editing
Does this mean that collaborative editing is being developped or even already implemented in quarto?

**Reply by cderv** [reply]
If you look at the code source reference for this deepwiki page, you'll find they refer to one of the component in the monorepo: https://github.com/quarto-dev/quarto/tree/main/packages/editor-codemirror

This component is for visual editor if I am not mistaken. 

We'll share updates regarding collaborative editing - don't worry. No need to ask tool like deepwiki for hidden feature 😃

**Comment by Acbeek** [comment]
Any new Quarto collaborative workflow suggestions?

**Reply by cderv** [reply]
We have ongoing work, and when this will be ready, there will be news to share and we will update and communicate.

No need to ping for update. Thank you
